### ISSUE TITLE
Best practices for auto-pagination / paginating in batches

### ISSUE THREAD

**Issue by kytrinyx** [octokit/discussions#3]
I think it would be lovely if each client used whatever iteration pattern that is the convention for the language, doing batching in the background (not fetching all the data up front).

**Comment by kytrinyx** [comment]
I'm pretty happy with how the google/go-github client does this. the `for { ... }` with a break is idiomatic in Go.

```
client := github.NewClient(nil)

opt := &github.RepositoryListByOrgOptions{
	ListOptions: github.ListOptions{PerPage: 10},
}

for {
	repos, resp, err := client.Repositories.ListByOrg(ctx, "github", opt)
	if err != nil {
		return err
	}

	// loop through repos here and do whatever you need to do

	if resp.NextPage == 0 {
		break
	}
	opt.Page = resp.NextPage
}
```

**Comment by kytrinyx** [comment]
Right now in the ruby client, paginating is pretty awkward. If you can use the `auto_paginate` fetaure, then it's OK, but as soon as your data sets are big then you use up all the memory and your app crashes.

Currently, if you find the documentation about paginating manually using `last_response.rels[:next].get`, then you typically end up having a thing where you need to process the first page of results separately from the results from all the other pages.

This might be a little bit of a problem with documentation, and I'm going to work with tarebyte on digging into that.

One of the things that I really like from ActiveRecord in Rails is the find in batches that hides the pagination/batching aspect for the user and lets you iterate through _all_ the records in a very ruby-ish way.

```
Thing.find_each do |thing|
  # handle each individual record however you like
end
```

**Comment by zeke** [comment]
Thanks for opening, @kytrinyx! I was going to do the same. :)

[ghutils](https://github.com/rvagg/ghutils) is "a collection of utility functions for dealing with the GitHub API". Its `lister` method is really convenient. Given a paginated url resource, it recursively fetches all available pages of data and returns an array containing the complete list. I'm not sure how it works under the hood (my guess is the requests are one at a time, serially) but I used it to consume [10,000 GitHub issues and their comments](https://github.com/zeke/electron-electron-issues#readme) and it worked like a charm with no rate limit issues.

**Comment by kytrinyx** [comment]
> fetches all available pages of data and returns an array containing the complete list.

This sounds like what the `auto_paginate` feature does in Ruby. I'll need to see if it does something differently in order to not run out of memory like the ruby one does.

**Comment by kytrinyx** [comment]
node-github also has that annoying "process it twice" thing going on that the ruby client has:

https://github.com/mikedeboer/node-github/blob/master/examples/paginationCustomHeaders.js#L19-L32

```
github.issues.getForRepo({
    owner: "mikedeboer",
    repo: "node-github",
    headers: customHeaders
}, function(err, res) {
    // process the results for the first page
    
    if (github.hasNextPage(res)) {
        github.getNextPage(res, customHeaders, function(err, res) {
            // process the results for the subsequent pages
        });
    }
});
```

**Comment by kytrinyx** [comment]
Oh, I just realized that the example that he has in the repo doesn't actually show how to paginate through all the results, just how to get the second page if there is one.

**Comment by bkeepers** [comment]
> Oh, I just realized that the example that he has in the repo doesn't actually show how to paginate through all the results, just how to get the second page if there is one.

Probot injects a [paginate method](https://github.com/probot/probot/blob/master/lib/paginate.js) on the node-github client, which takes a request as the first argument, and a callback as the second argument.

```js
github.paginate(github.issues.getAll(params), res => {
  // this callback is called for each page of results
  console.log('issues', res);
});
```

The callback will be invoked for each page of the results. The `paginate` method returns an array of all the values returned from the callback. If no callback is passed, then `paginate` just returns an array of results from all pages.

**Comment by kytrinyx** [comment]
I've been digging into the looping question a bit, and so far the thing that I have liked best as a sort of meta-pattern is this:

```
response = get the first page
do this forever {
  # process the response data

  if (there's no next page) {
    break out of the loop
  }
  response = get the next page
}
```

I really, really like the `github.paginate` thing, though. I wrote a similar thing in the Ruby using a while loop and yield so that we can give Ruby-folks a ruby-ish experience (because how often do you see an honest-to-goodness `while` in Ruby?)

**Comment by gr2m** [comment]
I did the the "send request until out of pages and then resolve with all results at once pattern` to in the past.

I’m not sure what the best approach would be for this. From a user perspective, without knowing the rate limit limitations of GitHub’s APIs, I would expect that `github.issues.getAll(params)` would resolve with all issues by default. Which means it would send multiple requests if necessary.

We could add another argument which would work similar to jQuery’s [.progress()](https://api.jquery.com/deferred.progress/) API, the passed callback would be called for each page result

```js
params.onPage = function (pageResult) {
  // called for each page result
}
github.issues.getAll(params)).then(res => {
  // called for all results from all pages
}
```

🤔

**Comment by kytrinyx** [comment]
> From a user perspective, without knowing the rate limit limitations of GitHub’s APIs, I would expect that github.issues.getAll(params) would resolve with all issues by default.

Yeah, that's a tricky one. It's hard to know what people expect, but rate limiting does seem to come as a surprise to people. I really don't know what the right approach is when rate limits are hit, and how best to give people the information they need at the right time.

**Comment by sigmavirus24** [comment]
It's unclear if non-octokit libraries are welcome in this discussion, but I'll add a thought. Python has the concept of generators, things that yield data and with that structure, it's easy to paginate all of the repositories on GitHub with [github3.py](/sigmavirus24/github3.py) without running out of memory (although the ratelimit will be a problem depending on what you do inside your loop.

That would look like

```py
import github3

gh = github3.login(...)
for repository in gh.all_repositories(...):
    # do something
```

The user doesn't have to leave the for-loop until they've exhausted all repositories they can see. In fact, all paginated resources work this way for github3.py. I feel like this wouldn't be terribly difficult for octokit.rb but it has been a while since I touched Ruby, so I could be misremembering some things.
